core/src/main/scala/scalaz/std/Map.scala

```diff
@@ -111,10 +111,11 @@ trait MapInstances extends MapInstances0 with MapFunctions {
     }
 
   implicit def mapShow[K, V](implicit K: Show[K], V: Show[V]): Show[Map[K, V]] =
-    Show.show(m => "Map[" +:
-                Cord.mkCord(", ", m.toSeq.map{
-                  case (k, v) => Cord(K show k, "->", V show v)
-                }: _*) :+ "]")
+    Show.show(m =>
+      Cord("Map[") ++
+        m.toList.foldLeft(Cord.empty) {
+          case (acc, (k, v)) => acc ++ K.show(k) ++ "->" ++ V.show(v) ++ ", "
+        } ++ Cord("]"))
```

Comment:
so, Map is a binary tree; you could do something like

```scala
Show.show { m =>
  def loop(b: Bin[K, V]) = b match {
    case Bin(k, v, l, r) =>
      val ls = l match {
        case Tip()        => Cord.Empty
        case l: Bin[K, V] => loop(b) ++ ", "
      }
      val rs = l match {
        case Tip()        => Cord.Empty
        case l: Bin[K, V] => ", " ++ loop(b)
      }
      ls ++ K.show(k) ++ "->" ++ V.show(v) ++ rs
  }
  m match {
    case Tip()        => "Map()"
    case b: Bin[K, V] => "Map(" ++ loop(b) ++ ")"
  }
}
```

Still not tailrec, but plausibly faster. (I didn't test it.)

Reply:
probably a lot of the data structures can have optimised `Show`s now...